vmm/src/device_manager.rs

```diff
@@ -2413,9 +2420,19 @@ impl DeviceManager {
             << 3;
 
         let memory = self.memory_manager.lock().unwrap().guest_memory();
+        #[allow(unused_variables)]
+        let vfio_container: Arc<VfioContainer>;
+        #[cfg(feature = "kvm")]
         let vfio_container = Arc::new(
             VfioContainer::new(device_fd.clone()).map_err(DeviceManagerError::VfioCreate)?,
         );
+        #[cfg(not(feature = "kvm"))]
+        return Err(DeviceManagerError::VfioCreate(
+            vfio_ioctls::VfioError::OpenContainer(io::Error::new(
+                io::ErrorKind::Other,
+                "KVM not supported",
+            )),
+        ));
```

Comment:
Do we really need that if `vm.create_passthrough_device()` returns an error for a non kvm hypervisor?

Reply:
Ah, I see why you're doing this...The `vfio-ioctls` crate does have a KVM dependency in practice, so we should simply not try to call into any VFIO API if we're not on KVM until we make `vfio-ioctls` hypervisor agnostic.

What about having something like:

```Rust
    fn add_passthrough_device(
        &mut self,
        pci: &mut PciBus,
        interrupt_manager: &Arc<dyn InterruptManager<GroupConfig = MsiIrqGroupConfig>>,
        device_fd: &Arc<DeviceFd>,
        device_cfg: &mut DeviceConfig,
    ) -> DeviceManagerResult<(u32, String)> {
        #[cfg(feature = "kvm")]
        return self.add_vfio_device(pci, interrupt_manager, device_fd, device_cfg);

        #[cfg(not(feature = "kvm"))]
        Err(DeviceManagerError::NoDevicePassthroughSupport)
    }

```

And replacing all `add_vfio_device()` calls with that one?

Reply:
I think this should work.